// @pnpm/network.proxy-agent/dist/proxy-agent.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const proxy_agent_1 = require("./proxy-agent");
const socks_proxy_agent_1 = require("socks-proxy-agent");
jest.mock('agentkeepalive', () => {
    const MockHttp = mockHttpAgent('http');
    MockHttp['HttpsAgent'] = mockHttpAgent('https');
    return MockHttp;
});
function mockHttpAgent(type) {
    return function Agent(opts) {
        return {
            ...opts,
            __type: type,
        };
    };
}
const OPTS = {
    agent: null,
    ca: 'ca',
    cert: 'cert',
    key: 'key',
    localAddress: 'localAddress',
    maxSockets: 5,
    strictSsl: true,
    timeout: 5,
};
test('all expected options passed down to proxy agent', () => {
    const opts = {
        httpsProxy: 'https://user:pass@my.proxy:1234/foo/',
        noProxy: 'qar.com, bar.com',
        ...OPTS,
    };
    const agent = (0, proxy_agent_1.getProxyAgent)('https://foo.com/bar', opts);
    expect(agent.connectOpts).toEqual({
        ALPNProtocols: ['http/1.1'],
        auth: 'user:pass',
        ca: 'ca',
        cert: 'cert',
        host: 'my.proxy',
        key: 'key',
        localAddress: 'localAddress',
        maxSockets: 5,
        port: 1234,
        rejectUnauthorized: true,
        timeout: 6,
    });
    expect(agent.proxy.protocol).toEqual('https:');
});
test('a socks proxy', () => {
    const opts = {
        httpsProxy: 'socks://user:pass@my.proxy:1234/foo',
        ...OPTS,
    };
    const agent = (0, proxy_agent_1.getProxyAgent)('https://foo.com/bar', opts);
    expect(agent instanceof socks_proxy_agent_1.SocksProxyAgent).toBeTruthy();
    expect(agent.proxy).toEqual({
        host: 'my.proxy',
        port: 1234,
        type: 5,
    });
});
test('proxy credentials are decoded', () => {
    const opts = {
        httpsProxy: `https://${encodeURIComponent('use@!r')}:${encodeURIComponent('p#as*s')}@my.proxy:1234/foo`,
        ...OPTS,
    };
    const agent = (0, proxy_agent_1.getProxyAgent)('https://foo.com/bar', opts);
    expect(agent.connectOpts).toEqual({
        ALPNProtocols: ['http/1.1'],
        auth: 'use@!r:p#as*s',
        ca: 'ca',
        cert: 'cert',
        host: 'my.proxy',
        key: 'key',
        localAddress: 'localAddress',
        maxSockets: 5,
        port: 1234,
        // protocol: 'https:',
        rejectUnauthorized: true,
        timeout: 6,
    });
    expect(agent.proxy.protocol).toEqual('https:');
});
test('proxy credentials are decoded', () => {
    const opts = {
        httpsProxy: 'https://use@!r:p#as*s@my.proxy:1234/foo',
        ...OPTS,
    };
    expect(() => (0, proxy_agent_1.getProxyAgent)('https://foo.com/bar', opts)).toThrow("Couldn't parse proxy URL");
});
